examples/jsm/csm/CSM.js

```diff
@@ -0,0 +1,328 @@
+
+			light.shadow.camera.updateMatrixWorld( true );
+
+			const bbox = new FrustumBoundingBox().fromFrustum( lightSpaceFrustum );
+			bbox.getSize();
+			bbox.getCenter( this.lightMargin );
+
+			const squaredBBWidth = Math.max( bbox.size.x, bbox.size.y );
+
+			let center = new THREE.Vector3( bbox.center.x, bbox.center.y, bbox.center.z );
+			center.applyMatrix4( light.shadow.camera.matrixWorld );
+
+			light.shadow.camera.left = - squaredBBWidth / 2;
+			light.shadow.camera.right = squaredBBWidth / 2;
+			light.shadow.camera.top = squaredBBWidth / 2;
+			light.shadow.camera.bottom = - squaredBBWidth / 2;
+
+			light.position.copy( center );
+			light.target.position.copy( center );
+
+			light.target.position.x += this.lightDirection.x;
+			light.target.position.y += this.lightDirection.y;
+			light.target.position.z += this.lightDirection.z;
+
+			light.shadow.camera.updateProjectionMatrix();
+			light.shadow.camera.updateMatrixWorld();
+
+		}
+
+	}
+
+	injectInclude() {
+
+		THREE.ShaderChunk.lights_fragment_begin = Shader.lights_fragment_begin;
+		THREE.ShaderChunk.lights_pars_begin = Shader.lights_pars_begin;
+
+	}
+
+	setupMaterial( material ) {
+
+		material.defines = material.defines || {};
+		material.defines.USE_CSM = 1;
+		material.defines.CSM_CASCADES = this.cascades;
+
+		const breaksVec2 = [];
+
+		for ( let i = 0; i < this.cascades; i ++ ) {
+
+			let amount = this.breaks[ i ];
+			let prev = this.breaks[ i - 1 ] || 0;
+			breaksVec2.push( new THREE.Vector2( prev, amount ) );
+
+		}
+
+		const self = this;
+
+		material.onBeforeCompile = function ( shader ) {
+
+			shader.uniforms.CSM_cascades = { value: breaksVec2 };
+			shader.uniforms.cameraNear = { value: self.camera.near };
+			shader.uniforms.shadowFar = { value: self.far };
+
+			self.materials.push( shader );
+
+		};
+
+	}
+
+	updateUniforms() {
+
+		for ( let i = 0; i < this.materials.length; i ++ ) {
+
+			this.materials[ i ].uniforms.CSM_cascades.value = this.getExtendedBreaks();
+			this.materials[ i ].uniforms.cameraNear.value = this.camera.near;
+			this.materials[ i ].uniforms.shadowFar.value = this.far;
+
+		}
+
+	}
+
+	getExtendedBreaks() {
+
+		let breaksVec2 = [];
+
+		for ( let i = 0; i < this.cascades; i ++ ) {
+
+			let amount = this.breaks[ i ];
+			let prev = this.breaks[ i - 1 ] || 0;
+			breaksVec2.push( new THREE.Vector2( prev, amount ) );
+
+		}
+
+		return breaksVec2;
+
+	}
+
+	setAspect( aspect ) {
+
+		this.aspect = aspect;
+		this.initCascades();
+
+	}
+
+	updateFrustums() {
+
+		this.getBreaks();
+		this.initCascades();
+		this.updateUniforms();
+
+	}
+
+	helper( cameraMatrix ) {
+
+		let frustum;
+		let geometry;
+		const material = new THREE.LineBasicMaterial( { color: 0xffffff } );
+		const object = new THREE.Object3D();
+
+		for ( let i = 0; i < this.frustums.length; i ++ ) {
+
+			frustum = this.frustums[ i ].toSpace( cameraMatrix );
+
+			geometry = new THREE.Geometry();
```

Comment:
Please avoid the usage of `Geometry` in this method. Always work with `BufferGeometry`.

Reply:
Fixed. I didn't really think about it when I was writing this. Helper was supposed to be used only for debugging.